Allow Plugins to request to perform cluster actions and index actions with their assigned PluginSubject and prompt on install

cwperks opened the pull request:
### Description

In https://github.com/opensearch-project/OpenSearch/pull/14630, a new extension point was created called IdentityAwarePlugin that has a single method called `assignSubject`. The subject that is given to IdentityAwarePlugin is intended to be a replacement for `try (ThreadContext.StoredContext ctx = threadContext.stashContext() { ... }`) and ensures that the security plugin can enforce authz checks on transport actions in this privileged block. The replacement is utilizing the assigned plugin subject and instead wrapping a block with `pluginSubject.runAs(() -> { ... })` which injects an identity associated with the plugin so that the Security plugin (i.e. the IdentityPlugin) can authorize the transport actions instead of allowing the plugin to operate unrestricted.

This PR allows a plugin to define a `plugin-permissions.yml` file on the same level as `plugin-security.policy` file that can have 2 keys: 1) `cluster.actions` and 2) `indices.actions`

Example

```
description: >
  This plugin needs permission to write to the security audit log index when configured
  to use an OpenSearch Index for the audit log. By default the index will match the pattern
  'security-auditlog-'YYYY.MM.dd, but can be configured with the plugins.security.audit.config.index
  setting
cluster.actions:
  - cluster:monitor/health
index.actions:
  security-auditlog*:
    - indices:data/write/index*
```

This PR modifies the `plugin-cli` to prompt a cluster admin with the requested permissions. See an example below where I modified the security plugin to be an `IdentityAwarePlugin` and request permission to write to the `security-auditlog-*` index pattern. The security plugin always needs to be able to write to this index pattern regardless of the authenticated user's permissions. 

```
➜  opensearch-3.0.0-SNAPSHOT git:(identity-aware-plugin-request-perms) ./bin/opensearch-plugin install file:/Users/cwperx/Projects/opensearch/security/build/distributions/opensearch-security-3.0.0.0-SNAPSHOT.zip
-> Installing file:/Users/cwperx/Projects/opensearch/security/build/distributions/opensearch-security-3.0.0.0-SNAPSHOT.zip
-> Downloading file:/Users/cwperx/Projects/opensearch/security/build/distributions/opensearch-security-3.0.0.0-SNAPSHOT.zip
[=================================================] 100%
Sep 05, 2024 4:45:01 PM org.apache.lucene.internal.vectorization.VectorizationProvider lookup
WARNING: Java vector incubator module is not readable. For optimal vector performance, pass '--add-modules jdk.incubator.vector' to enable Vector API.
16:45:01.758 [main] INFO  org.opensearch.security.ssl.transport.SSLConfig - SSL dual mode is disabled
16:45:01.759 [main] WARN  org.opensearch.security.OpenSearchSecurityPlugin - OpenSearch Security plugin installed but disabled. This can expose your configuration (including passwords) to the public.
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
@     WARNING: plugin requires additional permissions     @
@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@@
* java.io.FilePermission /proc/sys/net/core/somaxconn#plus read
* java.lang.RuntimePermission accessClassInPackage.com.sun.jndi.*
* java.lang.RuntimePermission accessClassInPackage.sun.misc
* java.lang.RuntimePermission accessClassInPackage.sun.nio.ch
* java.lang.RuntimePermission accessClassInPackage.sun.security.x509
* java.lang.RuntimePermission accessDeclaredMembers
* java.lang.RuntimePermission accessUserInformation
* java.lang.RuntimePermission createClassLoader
* java.lang.RuntimePermission getClassLoader
* java.lang.RuntimePermission setContextClassLoader
* java.lang.RuntimePermission shutdownHooks
* java.lang.reflect.ReflectPermission suppressAccessChecks
* java.net.NetPermission getNetworkInformation
* java.net.NetPermission getProxySelector
* java.net.SocketPermission * connect,accept,resolve
* java.security.SecurityPermission getProperty.org.bouncycastle.ec.max_f2m_field_size
* java.security.SecurityPermission getProperty.org.bouncycastle.pkcs12.default
* java.security.SecurityPermission getProperty.org.bouncycastle.rsa.max_mr_tests
* java.security.SecurityPermission getProperty.org.bouncycastle.rsa.max_size
* java.security.SecurityPermission getProperty.ssl.KeyManagerFactory.algorithm
* java.security.SecurityPermission insertProvider.BC
* java.security.SecurityPermission org.apache.xml.security.register
* java.security.SecurityPermission putProviderProperty.BC
* java.security.SecurityPermission removeProviderProperty.BC
* java.security.SecurityPermission setProperty.ocsp.enable
* java.util.PropertyPermission * read,write
* javax.security.auth.AuthPermission doAs
* javax.security.auth.AuthPermission modifyPrivateCredentials
* javax.security.auth.kerberos.ServicePermission * accept
See http://docs.oracle.com/javase/8/docs/technotes/guides/security/permissions.html
for descriptions of what these permissions allow and the associated risks.

Plugin requests permission to perform the following transport actions. Any index
pattern that appears below is a default value and may change depending on plugin settings.

Cluster Actions
---------------

* cluster:monitor/health

Index Actions
-------------

Index Pattern: security-auditlog-*

* indices:data/write/index

Continue with installation? [y/N]
```

In order to extract the requested actions in `plugin-cli` I had to create a new method in the PluginsService to try instantiating a plugin with empty settings.

### Related Issues

https://github.com/opensearch-project/OpenSearch/issues/15958

### Check List
- [ ] Functionality includes testing.
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

cwperks commented:
@sohami I opened this Draft PR to show how a mechanism can work for plugins to request to perform an enumerable set of actions within a `pluginSubject.runAs(() -> { ... })` block (replacement for `try (ThreadContext.StoredContext ctx = threadContext.stashContext()) { ... }`).

https://github.com/opensearch-project/opensearch-plugins/issues/238#issuecomment-2319034570'

@prudhvigodithi You may be interested in this PR as well.

reta reviewed (changes requested):
Thanks @cwperks , this feature definitely lacks the design document (in my opinion), hence the few things I would like to highlight:
 - our security  model is declarative  (policy file, yml file, ...), why we are changing that to be code driven?
 - what are the consequences of loading plugins at the installation time? we have no control over any initialization logic,  there are no security sandbox established so plugins could do literally anything at this point?
 - what is the relation between identity and plugin permissions? why are gluing this two together?

cwperks commented:
> Thanks @cwperks , this feature definitely lacks the design document (in my opinion), hence the few things I would like to highlight

Thank you for the initial comments @reta. I will open an issue that pertains to just the changes in this PR.

On a high-level, the changes in this PR are targeting 2 cases (so far) identified in plugins that need to perform transport actions outside of authenticated user context:

- The Security plugin needs to write to the security audit log index (when using `internal_opensearch` sink)
- Another plugin-dev chimed in [here](https://github.com/opensearch-project/opensearch-plugins/issues/238#issuecomment-2319034570) that they have a plugin that needs to perform node stats regardless of the authenticated users privileges

This PR seeks to give plugin devs a generic way to define actions that plugin can perform with its identity (hence tying is to IdentityAwarePlugin interface)

> our security model is declarative (policy file, yml file, ...), why we are changing that to be code driven (for permissions)?

The JSM permissions are policy-file driven and this PR does not change that. See the response to the question above, this change pertains to IdentityAwarePlugins so I thought it was appropriate to add to that interface. It could be done through a file (similar to plugin-security.policy file), but since it was applicable to a single type of plugin I chose to add to the interface and define it in code.

> what are the consequences of loading plugins at the installation time? we have no control over any initialization logic, there is no security sandbox established so plugins could do literally anything at this point?

If this approach is taken, then is it possible to prompt twice: 1) For JSM permissions, 2) for requested transport actions. On the second prompt is it possible to enforce the JSM permissions agreed to in the first prompt to ensure that loading a plugin is done with the same permissions that were just agreed upon?

If the code approach has too many issues, then I could look into a file-based approach similar to `plugin-security.policy`

> what is the relation between identity and plugin permissions? why we are gluing this two together?

Identity is what gives these transport action permissions functionality. While it could be done generically for any plugins, it would only be the IdentityAwarePlugins that get use out of these while authz is in the security plugin.

reta commented:
> Thank you for the initial comments @reta. I will open an issue that pertains to just the changes in this PR.

Please do, thank you.

cwperks commented:
This PR addresses a previous conversation we had here: https://github.com/opensearch-project/OpenSearch/issues/14733#issuecomment-2233153734

cwperks commented:
@reta opened an issue here to discuss: https://github.com/opensearch-project/OpenSearch/issues/15958

I tried to outline the problem to make it clear what problem this is trying to solve without getting too granular into various solutions.

cwperks commented on server/src/main/java/org/opensearch/plugins/PluginSecurity.java:
```diff
@@ -171,4 +224,12 @@ public static Set<String> parsePermissions(Path file, Path tmpDir) throws IOExce
+     */
+    @SuppressWarnings("removal")
+    public static Settings parseRequestedActions(Path file) throws IOException {
+        return Settings.builder().loadFromPath(file).build();
```
I chose to use the Settings object in here since Settings already supports reading in Yaml files.

cwperks commented:
@reta I'd like to bring this PR back up in the near future. I addressed the comments to read the requested perms from a file instead of having a plugin define them in code. This also changes `IdentityPlugin.getPluginSubject` extension point to accept a PluginInfo option instead of the full Plugin object.

reta commented on server/src/main/java/org/opensearch/plugins/IdentityPlugin.java:
```diff
@@ -38,8 +38,8 @@ public interface IdentityPlugin {
      * @return Subject corresponding to the plugin
      */
-    PluginSubject getPluginSubject(Plugin plugin);
+    PluginSubject getPluginSubject(PluginInfo pluginInfo);
```
Why do we need to use `PluginInfo`? Plugin does not need to be told what action he is allowed to use: it comes from the its permissions policy.

cwperks replied:
Its lighter-weight than Plugin and its the vehicle that carries the actions that a plugin may request to perform with its PluginSubject.

The security plugin would need these 2 pieces of information to form the PluginSubject:

1. Plugin Identifier (by convention it would be the canonical class name of the plugin)
2. Request Actions <- This is what this PR is introducing. i.e. The security plugin is requesting permissions to write to the security auditlog index with its pluginSubject.

reta replied:
Got it, this is indeed the conceptual problem (at least I overlooked). I think this pull request is a bit ahead of its time, here is my reasoning:
 - the core has no notion of permissions whatsoever
 - as of today, this is purely and only `security` plugin feature, and
 - any alternative (identity?) plugin implementation would have known nothing about permissions

I still think that the idea to expose the plugin permissions is sound but we need a way to formalize the permissions first and (may be?) make core / other plugins aware of them (basically, do what we have done with `Subject` etc). 

Why I think that is important, when we install any plugin right now, the security policy will be enforced no matter what. With permissions, as it stand now, this is purely hint, nothing will happen if `security` plugin is not installed / disabled (or any other its replacement is present).

@peternied sorry to dragging you here, but I think the subject of permissions what discussed before, right?

cwperks replied:
FWIW I think presenting this information to an administrator installing a plugin empowers them with more information about what actions a plugin will perform in the cluster regardless if the security plugin is installed or not.

I left a comment [here](https://github.com/opensearch-project/security/issues/4439#issuecomment-2174505670) about complications of putting system index protection directly in the core. The biggest challenge with that is that operators may need to be able to perform invasive actions on system indices and do so by presenting the admin certificate which is another feature of the security plugin.

cwperks replied:
JSM Permissions actually work the same way. A cluster operator is prompted with the permissions on install even if they run opensearch with JSM disabled.

reta replied:
> JSM Permissions actually work the same way. A cluster operator is prompted with the permissions on install even if they run opensearch with JSM disabled.

I believe it is not possible to disable ISM (except tests), it is installed unconditional (please correct me if I am wrong here).

reta replied:
> FWIW I think presenting this information to an administrator installing a plugin empowers them with more information about what actions a plugin will perform in the cluster regardless if the security plugin is installed or not.

This is right, and I agree with that. The question is how to do that in a way that if administrator says "Looks good" but nothing is enforced?

cwperks replied:
You're right its not possible to disable outside of the [tests](https://github.com/opensearch-project/OpenSearch/blob/main/test/framework/src/main/java/org/opensearch/bootstrap/BootstrapForTesting.java#L128). JSM being removed in JDK 24 is a disruptive change. lmk if I can help with review on any items.

reta replied:
Please join the discussion https://github.com/opensearch-project/OpenSearch/issues/1687

cwperks replied:
Hey @reta, I opened a draft PR in core to add a new sandbox module for system index protection: https://github.com/opensearch-project/OpenSearch/pull/16695

In its current state it is a crude implementation and would need to be expanded upon to match the functionality that the security plugin provides.

I wanted to open a draft to demonstrate some of the challenges in moving system index protection to the core. Ultimately, I think it is a good idea so that they are protected regardless if the security plugin is installed and enabled. 

One of the biggest challenges is generic ActionRequest -> indices() resolution in an action filter. The way this works in the security plugin is the [IndexResolverReplacer.getOrReplaceAllIndices()](https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/resolver/IndexResolverReplacer.java#L640-L832).

reta replied:
> Hey @reta, I opened a draft PR in core to add a new sandbox module for system index protection: #16695

Apologies for the delay @cwperks , I will try to find the time next week to look at it, thanks !

reta commented on server/src/main/java/org/opensearch/plugins/PluginInfo.java:
```diff
@@ -73,11 +76,22 @@ public class PluginInfo implements Writeable, ToXContentObject {
+        Function.identity()
+    );
+
+    public static final Setting<Settings> INDEX_ACTIONS_SETTING = Setting.groupSetting("index.actions.");
```
Oh, I think it comes from the security plugin [1], right? It certainly would make sense to keep it the same. Could we extract this logic somewhere (if it makes sense, for permissions only) so core and `security` would not duplicate it?

[1] https://github.com/opensearch-project/security/blob/main/config/roles.yml

cwperks replied:
Its not quite coming from the security plugin. The security plugin defines permissions like this: https://github.com/opensearch-project/security/blob/main/src/main/resources/static_config/static_roles.yml#L11-L17

```
all_access:
  static: true
  description: "Allow full access to all indices and all cluster APIs"
  cluster_permissions:
    - "*"
  index_permissions:
    - index_patterns:
        - "*"
      allowed_actions:
        - "*"
```

Cluster actions would be a list of action names, but index actions need a corresponding index pattern that the actions would be applicable to. For ultimate flexibility, different index patterns may have a different set of index actions that a plugin would be permitted to perform on the target index pattern.

cwperks replied:
Picking this back up, I will try addressing your second comment above [here](https://github.com/opensearch-project/OpenSearch/pull/15778#discussion_r1834705829).

I've been thinking through the [comment you posted on the other PR](https://github.com/opensearch-project/OpenSearch/pull/17296#issuecomment-2646619800), but I'm not quite sure how to address it. I agree that it would be better than calling `PluginSubject.runAs(() -> { ... })` but idk what the interface should look like.

cwperks replied:
The way the roles definition works in the security plugin is that it uses jackson and the corresponding CType (in this case RoleV7) to read a role from yaml.

1. Where the role is read in: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/securityconf/impl/SecurityDynamicConfiguration.java#L103
2. RoleV7 definition: https://github.com/opensearch-project/security/blob/main/src/main/java/org/opensearch/security/securityconf/impl/v7/RoleV7.java#L40-L47

cwperks replied:
How can security and core be unified here?

Core has some rudimentary action name validation in [TransportService](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/transport/TransportService.java#L1106-L1146)

nibix replied:
Just two random ideas crossing my mind. Basic assumption: Core does not do with this data anything but echoing it and passing it on to the security plugin, right?

- Option 1: So, why not just keep a raw AST of the config and pass that on to the security plugin?
- Option 2: Ideally (but likely not realistically due to effort constraints): During plugin installation, the installer tool would pass the config on to a security component (not the plugin, as there's no running node at that moment), which converts the config to a JWT - which in the end is just an opaque string. All the core needs to do is then passing that opaque string around.

nibix commented:
I figured that it might make sense to re-post a comment I wrote at https://github.com/opensearch-project/security/pull/4896#discussion_r1889683926

OpenSearch core already contains ground works for a protection of system indices. It works like this:

- In case a `REST` request is issues against a system index, the following message will be logged: `this request accesses system indices: [{}], but in a future major version, direct access to system indices will be prevented by default` 
- Exception: The HTTP header `X-opensearch-product-origin` is present in the REST request
- Other exception: The `RestHandler` implementation overrides `allowSystemIndexAccessByDefault()` and returns `true`.

See this commit for details: https://github.com/opensearch-project/OpenSearch/commit/5c8b0662df09a5d7a53fd0d98b1d37b1d831ddbc

These concepts still come from the ES times. If I understand it correctly, the goal was to replace the warning message by an error as a breaking change with a major release. 

IMHO, this concept makes totally sense for core. As @cwperks has pointed out, core has no safe means of authentication. What core can do is a kind of cooperative check: If a client assures that they need to perform operations on a system index, they are allowed to. If the client signals no concrete intent to operate on system index, they are not allowed to. This avoids unwanted operations in case of bugs on side of the client - be it an overly broad index pattern or a similar case. Of course, it does not provide real security. But that just takes into account the fact that core has no means to perform a secure authentication.

IMHO, a good strategy would be to build on the present ground works and to complete them. That would be also useful from a code hygiene point of view, as otherwise that existing code would be there without real purpose.

Relevant code pieces:

https://github.com/opensearch-project/OpenSearch/blob/57a660558f925da44eca18a7eb21983e61a7124a/server/src/main/java/org/opensearch/cluster/metadata/IndexNameExpressionResolver.java#L359-L378

https://github.com/opensearch-project/OpenSearch/blame/57a660558f925da44eca18a7eb21983e61a7124a/server/src/main/java/org/opensearch/rest/RestController.java#L374-L381

https://github.com/opensearch-project/OpenSearch/blob/57a660558f925da44eca18a7eb21983e61a7124a/server/src/main/java/org/opensearch/cluster/metadata/IndexNameExpressionResolver.java#L767-L774

dbwiddis commented on server/src/main/java/org/opensearch/plugins/PluginInfo.java:
```diff
@@ -561,6 +575,14 @@ public Map<String, List<String>> getIndexActions() {
+    /**
+     * Parses plugin-permissions.yml file.
+     */
+    @SuppressWarnings("removal")
```
I don't see anything in this method or the method it calls that's deprecated. Why this annotation?  Adding a new method with deprecated methods seems wrong... if anything is actually deprecated.

cwperks replied:
removed

dbwiddis reviewed (commented):
Not sure the state of this but it popped up as recently commented so I peeked at a few things.

cwperks commented on distribution/tools/plugin-cli/src/main/java/org/opensearch/tools/cli/plugin/PluginSecurity.java:
```diff
@@ -224,12 +223,4 @@ static Set<String> parsePermissions(Path file, Path tmpDir) throws IOException {
-    @SuppressWarnings("removal")
-    public static Settings parseRequestedActions(Path file) throws IOException {
-        return Settings.builder().loadFromPath(file).build();
-    }
```
Looks like there may have been rebasing problems. This is a brand new method.

cwperks commented on server/src/main/java/org/opensearch/plugins/PluginInfo.java:
```diff
@@ -234,13 +235,23 @@ public PluginInfo(final StreamInput in) throws IOException {
         }
     }
 
+    static boolean isOptionalExtension(String extendedPlugin) {
```
This code is already merged into main ([PR](https://github.com/opensearch-project/OpenSearch/pull/16909)) and not related to this PR. Is this showing up when browsing a single commit?
